### ISSUE TITLE
Python 3.7 can cast object to complex type

### ISSUE THREAD

**Issue by bear24rw** [numpy/numpy#11993]
### Reproducing code example:

```python
import numpy as np
print(np.__version__)
arr = np.array(['AAAAA', 18465886.0, 18465886.0], dtype=object)
print(arr.astype(np.complex64))
print(arr.astype(np.complex64)) 
```

### Error message:
Python3.6 works as I would expect (complains that it cannot convert a string to complex dtype)
```
$ python3.6 test.py 
1.16.0.dev0+7fcba41
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    print(arr.astype(np.complex64))
TypeError: must be real number, not str
```
Python3.7 converts it to a complex type and then errors if you try to do it again
```
$ python3.7 test.py 
1.16.0.dev0+7fcba41
[-6.3090155e-31+4.558e-41j  1.8465886e+07+0.000e+00j
  1.8465886e+07+0.000e+00j]
Traceback (most recent call last):
  File "test.py", line 5, in <module>
    print(arr.astype(np.complex64)) 
TypeError: must be real number, not str
```
### Numpy/Python version information:
```
>>> import sys
>>> sys.version
'3.7.0b3 (default, Mar 30 2018, 04:35:22) \n[GCC 7.3.0]'
>>> import numpy as np
>>> np.__version__
'1.16.0.dev0+7fcba41'
```

**Comment by eric-wieser** [comment]
Do you get this for `np.complex32` too?

I'm suspicious this might relate to #11998

**Comment by jjhelmus** [comment]
I was able to replicate this will various complex data types including `np.complex64`, `np.complex128` and `np.complex256`.

**Comment by hhuuggoo** [comment]
I don't think it is related to #11998.  The reason I say that is because I have 2 environments (py36 and numpy 1.15.2) and (py37 and numpy 1.15.2)

in both I get 
```
>>> (issubclass(np.float_, float),  issubclass(np.float_, np.floating))
(True, True)
>>> (issubclass(np.complex_, float),  issubclass(np.complex_, np.complexfloating))
(False, True)
```

**Comment by ahaldane** [comment]
Should be fixed by #12062. It was a missing error check, which caused the error to be erased by subsequence cast operations. That's why
```python
>>> np.array(['AAAAA', 18465886.0, 18465886.0], dtype=object).astype('c8')
```
fails to raise an error but 
```python
>>> np.array(['AAAAA', 'BBBB'], dtype=object).astype('c8')
```
correctly raises: Only the last processed array element could raise an error.
